### ISSUE TITLE
GIT LFS tracking old data

### ISSUE THREAD

**Issue by nullB0t** [git-lfs/git-lfs#1991]
I created a question on StackOverflow about this [here](https://stackoverflow.com/questions/42521702/git-lfs-tracking-old-data)

Basically I have a project with LFS tracking some files and I refactored and moved those files around, then my LFS tracking got too big for GitHub (surpassing the 1.5GB) and I noticed that I only have ~100MB of data for LFS to track, I deleted my remote and tried to clean up using BFG and re-tracked all my files but I still get an error while trying to push:

`Trying to push "Content/Static/Geometry/Meshes/1M_Cube.uasset" with OID f471a994148db424160e6589509afbe29b9205f5db07ab4c231811aefccff955.
Not found in D:\Projects\UnrealProjects\MasterOfShadows\.git\lfs\objects\f4\71.` (which is indeed empty, it should be tracked now in a different location)

Looks like my asset is being tracked with an old OID, when I run `git lfs ls-files` it shows me `8b84072d8a * Content/Static/Geometry/Meshes/1M_Cube.uasset` for that asset so it should now be stored under `D:\Projects\UnrealProjects\MasterOfShadows\.git\lfs\objects\8b\84.` (which is indeed populated with data)

Can you help me understand how to clean the LFS tracking so that it only takes my current `head`?

**Comment by technoweenie** [comment]
That "Trying to push" message means you have an old version of that file somewhere in one of the branches in the push to your git host. Try `git log Content/Static/Geometry/Meshes/1M_Cube.uasset` to see the old versions.

Two things to keep in mind here though:

1. Running `git log <file>` will only scan the history of your current branch. You can get a list of branches with `git branch`, and get the history of that file by running `git log <branch> Content/Static/Geometry/Meshes/1M_Cube.uasset`.
2. `git lfs ls-files` only shows the current _commit_, and no older versions of files.

If your local repository has branches that you don't want to push, you can specify with a command like `git push origin master` instead of just `git push --all`.

I don't have much BFG experience, and don't know what commands you ran, but this looks more like a case of BFG not writing the LFS objects to `.git/lfs/objects`. Have you tried https://github.com/bozaro/git-lfs-migrate? It is our [current recommendation for migrating to LFS](https://github.com/git-lfs/git-lfs/wiki/Tutorial#migrating-existing-repository-data-to-lfs).

**Comment by nullB0t** [comment]
I haven't tried that, I will look into it. My project is still on the master branch, I don't have anything else to push.

So my question is, let's forget about BFG, I have those old objects being tracked in the `.git/lfs/objects` that I want to clean and only track the ones on the `head` of my master, do you think that using the migration tool will help me get rid of those or should I create a new repository using my current `head` and track only those files?

Thanks for your help ^^

**Comment by technoweenie** [comment]
> I have those old objects being tracked in the .git/lfs/objects that I want to clean and only track the ones on the head of my master, do you think that using the migration tool will help me get rid of those or should I create a new repository using my current head and track only those files?

The migration tool should be moving the historical files to `.git/objects`. Right now, since Git keeps all old revisions as just LFS pointers, Git LFS expects to be able to find them in `.git/lfs/objects` or download them from your LFS server. Removing the history and just pushing a single new commit is an option for now, but I don't think you'd want to do that regularly.

Do you have a desire to drop ancient versions of your large files? Are you looking for a policy like "keep the most recent 10 versions of *.uasset files" or "keep any *.uasset files newer than 3 months"? This is something we discussed in https://github.com/git-lfs/git-lfs/issues/1101#issuecomment-203635949, but haven't started fleshing out yet.

**Comment by nullB0t** [comment]
>Do you have a desire to drop ancient versions of your large files? Are you looking for a policy like "keep the most recent 10 versions of *.uasset files" or "keep any *.uasset files newer than 3 months"? This is something we discussed in #1101 (comment), but haven't started fleshing out yet.

Yes this is indeed something that would be very useful, just imagine that by just moving those LFS files around in the project will increase the size of the remote, it would be even better if the files are just moved around instead of being copied so if you reset to a previous commit you only need to track the files that have been removed and the other ones will simply change it's location (unless they had changes)

**Comment by jomark-t** [comment]
I have a question, do you need to use git lfs prior to git lfs migrate? Does this command push the objects to LFS server?

**Comment by ttaylorr** [comment]
> do you need to use git lfs prior to git lfs migrate?

You do not have to use Git LFS prior to running the `git lfs migrate` command. This command is designed to take repositories that _aren't_ using Git LFS and configure them correctly to do so. The only requirement is that Git LFS is downloaded, and installed correctly (with `git lfs install`).

> Does this command push the objects to LFS server?

The `git lfs migrate` command does not push new LFS objects to the server. Any new Git LFS objects created locally during a migration will be pushed to the server when you run `git push`.
